internal/provider/atlas_migration_resource.go

```diff
@@ -0,0 +1,320 @@
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+// Delete implements resource.Resource.
+func (r *MigrationResource) Delete(ctx context.Context, req resource.DeleteRequest, resp *resource.DeleteResponse) {
+	// No-op, because we don't want to delete the database
+}
+
+// Validate implements resource.ResourceWithValidateConfig.
+func (r MigrationResource) ValidateConfig(ctx context.Context, req resource.ValidateConfigRequest, resp *resource.ValidateConfigResponse) {
+	var data MigrationResourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if data.Version.Value == "" {
+		resp.Diagnostics.AddAttributeWarning(
+			path.Root("version"),
+			"version is unset",
+			"We recommend that you use 'version' to specify a version of the migration to run.\n"+
+				"If you don't specify a version, the latest version will be used when the resource being created.\n"+
+				"For keeping the database schema up to date, you should use set the version to using the value from "+
+				"`atlas_migrate_status.next` or `atlas_migrate_status.latest`\n",
+		)
+	}
+}
+
+// ModifyPlan implements resource.ResourceWithModifyPlan.
+func (r *MigrationResource) ModifyPlan(ctx context.Context, req resource.ModifyPlanRequest, resp *resource.ModifyPlanResponse) {
+	var plan *MigrationResourceModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	var state *MigrationResourceModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+	if plan != nil {
+		report, err := r.client.Status(ctx, &atlas.StatusParams{
+			DirURL:          plan.DirURL.Value,
+			URL:             plan.URL.Value,
+			RevisionsSchema: plan.RevisionsSchema.Value,
+		})
+		if err != nil {
+			resp.Diagnostics.Append(atlas.ErrorDiagnostic(err, "Failed to read migration status"))
+			return
+		}
+		if plan.Version.IsNull() {
+			v := report.LatestVersion()
+			plan.Version = types.String{Value: v, Null: v == ""}
+		}
+		// TODO(giautm): Add Atlas-Lint to report the issues
+	}
+}
+
+func (r *MigrationResource) migrate(ctx context.Context, data *MigrationResourceModel) (diags diag.Diagnostics) {
+	statusReport, err := r.client.Status(ctx, &atlas.StatusParams{
+		DirURL:          data.DirURL.Value,
+		URL:             data.URL.Value,
+		RevisionsSchema: data.RevisionsSchema.Value,
+	})
+	if err != nil {
+		diags.Append(atlas.ErrorDiagnostic(err, "Failed to read migration status"))
+		return
+	}
+	// The amount of migrations to apply
+	// We calculate this by subtracting the version from the latest version
+	var amount = 0
+	if data.Version.IsNull() {
+		amount = len(statusReport.Pending)
+	} else {
+		for idx, v := range statusReport.Pending {
+			if v.Version == data.Version.Value {
+				amount = idx + 1
+				break
+			}
+		}
+		if amount == 0 {
```

Comment:
we shouldn't fail if the latest migration is already applied be a previous `terraform apply`